### PR TITLE
Changed readdir to readdir_r for thread-safety

### PR DESCRIPTION
**Pull request by Andargor**
My threaded app was crashing, and this was too nice to just flush. Hopefully no boo-boos! Compiles on both Windows and ubuntu/g++ 4.9.

A.

### CODE REVIEW

File: tinydir.h [hunk 1]
```diff
@@ -282,6 +289,12 @@ void tinydir_close(tinydir_dir *dir)
 _TINYDIR_FUNC
 int tinydir_next(tinydir_dir *dir)
 {
+#ifndef _MSC_VER
+	int error;
+	struct dirent entry;
+	/*struct dirent *result;*/
```
Comment (wjwwood) [inline tinydir.h]: Should this be left in?

**Comment by cxong** [comment]
Thanks, but I'm not sure this is safe. From the [manual](http://linux.die.net/man/3/readdir_r), it looks like `readdir_r` is reentrant because the resulting `struct dirent *` is pointing to a user-allocated `entry`, but if that is going out of scope then the pointed location could become invalid. `entry` will need to be part of `tinydir_dir`.

Also `entry` may need to be dynamically allocated; see the end of the manual.

File: tinydir.h (hunk 1 above)
Comment (Andargor) [inline tinydir.h]: @probably not :)

**Comment by Andargor** [comment]
The dynamic allocation may be problematic with the C syntax. I'm proposing a static allocation that should not go out of scope considering how tinydir_dir is used.

**Comment by Andargor** [comment]
Aw damn, that was bad. let me fix it

**Comment by Andargor** [comment]
There. The only question is if theres a situation where the tinydir_dir may be accessed by multiple threads.

**Comment by cxong** [comment]
Thanks, this is almost there. I think `_ep` should be dynamically allocated as the manual suggests, because otherwise with the right file system and long enough filename, a buffer overflow could occur. See this blog post for more details: [//elliotth.blogspot.com/2012/10/how-not-to-use-readdirr3.html](//elliotth.blogspot.com/2012/10/how-not-to-use-readdirr3.html)

I'm not sure what you mean by the dynamic allocation being problematic; @dariomanesku added the `_TINYDIR_MALLOC`/`_TINYDIR_FREE` macros [here](https://github.com/cxong/tinydir/commit/43d73602fe190917c2574755d1434aac0e0a524a) so you can just use them in `tinydir_open`/`tinydir_close` respectively.

So that just leaves the question of how to calculate the size of the buffer. Fortunately there's a nice post (with working code) you can copy from here: [https://womble.decadent.org.uk/readdir_r-advisory.html](https://womble.decadent.org.uk/readdir_r-advisory.html). The same page says that the problem only affects weird OSs that nobody uses like Solaris and BeOS, which is nice.

On that note, the buffer size calculation method could also be used for `tinydir_dir.name`, getting rid of `_TINYDIR_FILENAME_MAX`.

**Comment by Andargor** [comment]
I used your recommendations. There is segfault potential if someone calls tinydir_next without calling tinydir_open (dir->_ep may not be nulled). Also, error reporting could be better, just returning -1 for now.

**Comment by Andargor** [comment]
Hmm. checks are failing. Compiles fine here, although I did not use all headers from Ben Hitchings' code:

#include <sys/types.h>
#include <dirent.h>
#include <limits.h>
#include <stddef.h>
#include <unistd.h>

**Comment by Andargor** [comment]
Seems to have worked. I hope it's not too messy.

**Comment by cxong** [comment]
Cool! :sunglasses: I'll be able to merge it soon, just need to try it out. Also there's a few fixes to add but I can do that. As for errors, I think tinydir itself will only ever return 0 or -1, and if the user needs to know the reason they check `errno`. Most standard library functions (like `malloc`) do that for you, but I think there are a few places where tinydir explicitly sets `errno`.

**Comment by cxong** [comment]
By the way, can I add your name to the copyright header, and what name can I use? "Andargor" or some other name?

**Comment by Andargor** [comment]
Sorry for the late response. Of course, I added the code snippet author's name, if you wish to reformat this go ahead.

**Comment by Andargor** [comment]
You can use Andargor <andargor@yahoo.com>

**Pull request merged**